fix(worker): avoid a race in shutdown hooks

mjameswh opened the pull request:
## What was changed
In `Runtime`, deny adding more shutdown hooks once a shutdown signal has been received.

## Why?
Avoid a race situation where a process might be sent a shutdown signal, but some worker has not yet registered its shutdown hook, causing the worker to never terminate.

## Checklist
1. Closes #909

Sushisource commented on packages/worker/src/runtime.ts:
```diff
@@ -380,8 +381,13 @@ export class Runtime {
    */
-  public registerShutdownSignalCallback(callback: () => void): void {
-    this.shutdownSignalCallbacks.add(callback);
+  public registerShutdownSignalCallback(callback: () => void): boolean {
```
Docstring should be updated w/ meaning of return value

Sushisource reviewed (approved):
Makes sense to me

bergundy commented on packages/test/src/test-worker-lifecycle.ts:
```diff
@@ -5,45 +5,98 @@
+      t.is(worker.getState(), 'STOPPED');
+      await t.throwsAsync(worker.run(), { message: 'Poller was already started' });
+    } finally {
+      // Make sure Runtime is reset after this test
+      if (Runtime._instance) await Runtime._instance.shutdown();
+      Runtime.instance();
```
Why create a new instance in this test?
Shouldn't the runtime automatically be shutdown after this test?

mjameswh replied:
Actually, it is required only for the two mock tests. It automatically reset after native tests.

The thing is that the mock native worker doesn't register itself with Runtime, and obviously doesn't unregister either. Consequently, `Runtime.shutdownIfIdle()` doesn't gets called at the end of the mock tests, and therefore, the Runtime instance never get discarded.

bergundy commented on packages/worker/src/worker.ts:
```diff
@@ -1596,10 +1596,16 @@ export class Worker {
-    Runtime.instance().registerShutdownSignalCallback(shutdownCallback);
-
     try {
+      const shutdownCallback = () => this.shutdown();
```
There's some more teardown happening below that's missed here.

Note that below the worker transitions to "STOPPED" state before it finalizes shutdown.
We could change that so we transition only after finalizeShutdown is called below and reuse the same teardown code.

bergundy commented on packages/worker/src/runtime.ts:
```diff
@@ -377,11 +378,20 @@ export class Runtime {
    */
-  public registerShutdownSignalCallback(callback: () => void): void {
-    this.shutdownSignalCallbacks.add(callback);
+  public registerShutdownSignalCallback(callback: () => void): boolean {
```
What's the benefit of eagerly returning `false` here?

bergundy replied:
It seems like it would save some complexity on the worker side to always use a callback.

mjameswh commented:
@bergundy Pushed a second attempt, addressing most of your comments.

Note that at this point, I have kept the `Runtime._instance.shutdown()` calls at the end of the two mock workers shutdown tests. Removing these would be possible but would require that `MockNativeWorker::finalizeShutdown` actually do something on `Runtime`, for example calling `shutdownIfIdle()`. I don't feel like this is a good thing.

bergundy commented on packages/worker/src/runtime.ts:
```diff
@@ -101,6 +101,7 @@ export class Runtime {
   protected readonly logPollPromise: Promise<void>;
   public readonly logger: Logger;
   protected readonly shutdownSignalCallbacks = new Set<() => void>();
+  protected state: 'running' | 'shuttingdown' = 'running';
```
nit:

```suggestion
  protected state: 'RUNNING' | 'SHUTTING_DOWN' = 'RUNNING';
```

bergundy reviewed (approved):
LGTM, had a single nit.
Thanks!
